src/Orleans.Runtime/Placement/ResourceOptimizedPlacementDirector.cs

```diff
@@ -0,0 +1,321 @@
+                if (_siloStatistics.TryGetValue(silo, out var stats))
+                {
+                    var filteredStats = stats.Value;
+                    if (!filteredStats.IsOverloaded)
+                    {
+                        relevantSilos[relevantSilosCount++] = new(i, filteredStats);
+                    }
+                }
+            }
+
+            // Limit to the number of candidates added.
+            relevantSilos = relevantSilos[0..relevantSilosCount];
+            Debug.Assert(relevantSilos.Length == relevantSilosCount);
+
+            // Pick K silos from the list of compatible silos, where K is equal to the square root of the number of silos.
+            // Eg, from 10 silos, we choose from 4.
+            int candidateCount = (int)Math.Ceiling(Math.Sqrt(relevantSilosCount));
+            ShufflePrefix(relevantSilos, candidateCount);
+            var candidates = relevantSilos[0..candidateCount];
+
+            (int Index, float Score) pick = (0, 1f);
+
+            foreach (var (index, statistics) in candidates)
+            {
+                float score = CalculateScore(statistics);
+
+                // It's very unlikely, but there could be more than 1 silo that has the same score,
+                // so we apply some jittering to avoid pick the first one in the short-list.
+                float scoreJitter = Random.Shared.NextSingle() / 100_000f;
+
+                if (score + scoreJitter < pick.Score)
+                {
+                    pick = (index, score);
+                }
+            }
+
+            return pick;
+        }
+
+        // Variant of the Modern Fisher-Yates shuffle which stops after shuffling the first `prefixLength` elements,
+        // which are the only elements we are interested in.
+        // See: https://en.wikipedia.org/wiki/Fisher%E2%80%93Yates_shuffle
+        static void ShufflePrefix(Span<(int SiloIndex, ResourceStatistics SiloStatistics)> values, int prefixLength)
+        {
+            Debug.Assert(prefixLength >= 0);
+            Debug.Assert(prefixLength <= values.Length);
+            var max = values.Length;
+            for (var i = 0; i < prefixLength; i++)
+            {
+                var chosen = Random.Shared.Next(i, max);
+                if (chosen != i)
+                {
+                    (values[chosen], values[i]) = (values[i], values[chosen]);
+                }
+            }
+        }
+    }
+
+    private bool IsLocalSiloPreferable(IPlacementContext context, SiloAddress[] compatibleSilos, float bestCandidateScore)
+    {
+        if (context.LocalSiloStatus != SiloStatus.Active || !compatibleSilos.Contains(context.LocalSilo))
+        {
+            return false;
+        }
+
+        if (_siloStatistics.TryGetValue(context.LocalSilo, out var localStats))
+        {
+            float localScore = CalculateScore(localStats.Value);
+            float scoreDiff = Math.Abs(localScore - bestCandidateScore);
```

Comment:
I believe this is incorrect. Consider:
`localScore = 0` (best possible score)
`bestCandidateScore = .2` (by chance, the local silo was not examined, so it is not in the candidate set)
`LocalSiloPreferenceMargin = 0.1`

We *should* preference the local silo here, since it's better even without the margin, but we won't because the condition below equates to `false` (i.e, `0.1 >= 0.2` is `false`).

Instead, the logic should be: `return localScore - _options.LocalSiloPreferenceMargin <= bestCandidateScore;`
That way, `LocalSiloPreferenceMargin` is strictly a boost to the local silo score.